BootloaderCorePkg/Library/FspApiLib/FspMemoryInit.c

```diff
@@ -91,3 +114,110 @@ CallFspMemoryInit (
+      !EFI_ERROR(Status) &&
+      !(Status & ENCODE_RESET_REQUEST(0));
+    GetNumPhasesParams.PhasesExecuted++, MultiPhaseInitParams.PhaseIndex++)
+  {
+    Status = CallFspMultiPhaseMemoryInit(&MultiPhaseInitParams);
+    ASSERT_EFI_ERROR(Status);
+    Status = FspVariableHandler(Status,CallFspMultiPhaseMemoryInit);
+    ASSERT_EFI_ERROR(Status);
+  }
+  return Status;
+}
+
+/**
+  This FSP API provides multi-phase memory and silicon initialization, which brings
+  greater modularity to the existing FspMemoryInit() and FspSiliconInit() API. Increased
+  modularity is achieved by adding an extra API to FSP-M and FSP-S. This allows the
+  bootloader to add board specific initialization steps throughout the MemoryInit and
+  SiliconInit flows as needed. The FspMemoryInit() API is always called before
+  FspMultiPhaseMemInit(); it is the first phase of memory initialization. Similarly, the
+  FspSiliconInit() API is always called before FspMultiPhaseSiInit(); it is the first phase of
+  silicon initialization. After the first phase, subsequent phases are invoked by calling the
+  FspMultiPhaseMem/SiInit() API.
+  The FspMultiPhaseMemInit() API may only be called after the FspMemoryInit() API and
+  before the FspSiliconInit() API; or in the case that FSP-T is being used, before the
+  TempRamExit() API. The FspMultiPhaseSiInit() API may only be called after the
+  FspSiliconInit() API and before NotifyPhase() API; or in the case that FSP-I is being used,
+  before the FspSmmInit() API. The multi-phase APIs may not be called at any other time.
+
+  @param[in] MultiPhaseInitParamPtr   Pointer to provide multi-phase init parameters.
+
+  @retval EFI_SUCCESS                 FSP execution environment was initialized successfully.
+  @retval EFI_INVALID_PARAMETER       Input parameters are invalid.
+  @retval EFI_UNSUPPORTED             The FSP calling conditions were not met.
+  @retval EFI_DEVICE_ERROR            FSP initialization failed.
+  @retval EFI_OUT_OF_RESOURCES        Stack range requested by FSP is not met.
+  @retval FSP_STATUS_RESET_REQUIREDx  A reset is reuired. These status codes will not be returned during S3.
+  @retval FSP_STATUS_VARIABLE_REQUEST An FSP variable access is required.
+**/
+EFI_STATUS
+EFIAPI
+CallFspMultiPhaseMemoryInit (
+  FSP_MULTI_PHASE_PARAMS     *MultiPhaseInitParamPtr
+  )
+{
+  FSP_INFO_HEADER             *FspHeader;
+  FSP_MULTI_PHASE_MEM_INIT    FspMultiPhaseMemoryInit;
+  EFI_STATUS                  Status;
+  UINT32                      FspmBase;
+
+  FspmBase = PCD_GET32_WITH_ADJUST (PcdFSPMBase);
+
+  FspHeader = (FSP_INFO_HEADER *)(UINTN)(FspmBase + FSP_INFO_HEADER_OFF);
+
+  ASSERT (FspHeader->Signature == FSP_INFO_HEADER_SIGNATURE);
+  ASSERT (FspHeader->ImageBase == FspmBase);
+
+  if (FspHeader->HeaderRevision < FSP24_HEADER_REVISION) {
+    return EFI_UNSUPPORTED;
+  }
+
+  ASSERT (FspHeader->FspMultiPhaseMemInitEntryOffset != 0);
```

Comment:
In case Multi-Phase memory API is not available in FSP, this assert fails. Can this assert be replaced with if condition checking whether FspMultiPhaseMemInitEntryOffset is equal to 0 and return EFI_UNSUPPORTED in that case?